File: backend/main/views/revenues.go
package views

import (
	"database/sql"
	"frostbox/contract"

	"github.com/shopspring/decimal"
)

type RevenueViewer struct {
	db *sql.DB
}

func NewRevenueViewer(db *sql.DB) *RevenueViewer {
	return &RevenueViewer{db: db}
}
func (r *RevenueViewer) GetRevenues() (contract.GetRevenuesResponse, error) {
	query := `SELECT 
		r.id, 
		r.value_main_currency, 
		r.vat_value_main_currency, 
		r.vat_rate, 
		r.title, 
		r.currency, 
		r.invoice_date, 
		r.revenue_date, 
		r.created_at,
		v.registration_number,
		d.readable_id
	FROM 
		revenues r
	LEFT JOIN 
		vehicles v ON r.vehicle_id = v.id
	LEFT JOIN 
		documents d ON r.document_id = d.id;`

	rows, err := r.db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var resp contract.GetRevenuesResponse
	for rows.Next() {
		var (
			valueMainCurrency, vatValueMainCurrency, vatRate float64
			revenue                                          contract.RevenuesResponse
		)

		if err := rows.Scan(
			&revenue.ID,
			&valueMainCurrency,
			&vatValueMainCurrency,
			&vatRate,
			&revenue.Title,
			&revenue.Currency,
			&revenue.InvoiceDate,
			&revenue.RevenueDate,
			&revenue.CreatedAt,
			&revenue.VehicleRegistrationNumber,
			&revenue.DocumentID,
		); err != nil {
			return nil, err
		}

		revenue.Value = roundUp(decimal.NewFromFloat(valueMainCurrency))
		revenue.VatValue = roundUp(decimal.NewFromFloat(vatValueMainCurrency))
		revenue.VatRate = decimal.NewFromFloat(vatRate).String()

		resp = append(resp, revenue)
	}

	return resp, nil
}

File: backend/uta-processor/app/revenues.py
from sqlalchemy import text

def add_revenue(
    session,
    value_main_currency,
    vat_value_main_currency,
    value,
    registration_number,
    vat_rate,
    currency,
    vat_value,
    invoice_date,
    revenue_date,
    title,
    document_id
):
    session.execute(
        text(
            """
            INSERT INTO revenues (
                value_main_currency,
                vat_value_main_currency,
                value,
                vehicle_id,
                vat_rate,
                currency,
                vat_value,
                invoice_date,
                revenue_date,
                title,
                document_id
            )
            VALUES (
                :value_main_currency,
                :vat_value_main_currency,
                :value,
                (SELECT id FROM vehicles WHERE registration_number = :registration_number LIMIT 1),
                :vat_rate,
                :currency,
                :vat_value,
                :invoice_date,
                :revenue_date,
                :title,
                :document_id
            )
            """
        ),
        {
            "value_main_currency": value_main_currency,
            "vat_value_main_currency": vat_value_main_currency,
            "value": value,
            "registration_number": registration_number,
            "vat_rate": vat_rate,
            "currency": currency,
            "vat_value": vat_value,
            "invoice_date": invoice_date,
            "revenue_date": revenue_date,
            "title": title,
            "document_id": document_id,
        },
    )
